File: 16_02_2024_Flatten_BST_to_sorted_list.cpp

class Solution
{
public:
    Node *flattenBST(Node *root)
    {
        // code here
        
        vector<Node*> ans;
        func(ans, root);
        Node* head = nullptr, *prev = nullptr;
        if (!ans.empty()) {
            head = ans[0];
            head->left = nullptr;
            prev = head;
            for (int i = 1; i < ans.size(); i++) {
                prev->right = ans[i];
                ans[i]->left = NULL;
                prev = ans[i];
            }
        }
        return head;
    }
    void func(vector<Node*> &ans, Node* root){
        if(!root) return;
        func(ans,root->left);
        ans.push_back(root);
        func(ans,root->right);
    }
};

File: 16_02_2024_Flatten_BST_to_sorted_list.java
class Pair{
    Node head;
    Node tail;
    Pair(Node h,Node t){
        head = h;
        tail = t;
    }
}
class Solution {
    private Pair helper(Node root){
        if(root==null)
            return new Pair(null,null);
        
        Pair leftP = helper(root.left);
        Pair rightP = helper(root.right);
        
        Pair ans = new Pair(root,root);
        if(leftP.head!=null){
            leftP.tail.right = root;
            ans.head = leftP.head;
        }
        if(rightP.tail!=null){
            root.right = rightP.head;
            ans.tail = rightP.tail;
        }
        root.left=null;
        return ans;
    }
    public Node flattenBST(Node root) {
        // Code here
        Pair ans = helper(root);
        return ans.head;
    }
}
